app/abci/handlers.go

```diff
@@ -0,0 +1,270 @@
+// a canonical set of vote extensions in the proposal.
+func (h *Handlers) PrepareProposalHandler() sdk.PrepareProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+		h.logger.Debug("start prepare proposal handler")
+
+		var injection []byte
+		if req.Height > ctx.ConsensusParams().Abci.VoteExtensionsEnableHeight {
+			err := baseapp.ValidateVoteExtensions(ctx, h.stakingKeeper, req.Height, ctx.ChainID(), req.LocalLastCommit)
+			if err != nil {
+				return nil, err
+			}
+
+			injection, err = json.Marshal(req.LocalLastCommit)
+			if err != nil {
+				h.logger.Error("failed to marshal extended votes", "err", err)
+			}
+
+			injectionSize := int64(len(injection))
+			if injectionSize <= req.MaxTxBytes {
+				req.MaxTxBytes -= injectionSize
+			} else {
+				h.logger.Error(
+					"vote extension size exceeds block size limit",
+					"injection_size", injectionSize,
+					"MaxTxBytes", req.MaxTxBytes,
+				)
+				return &abci.ResponsePrepareProposal{Txs: make([][]byte, 0)}, ErrVoteExtensionInjectionTooBig
+			}
+		}
+
+		defaultRes, err := h.defaultPrepareProposal(ctx, req)
+		if err != nil {
+			h.logger.Error("failed to run default prepare proposal handler", "err", err)
+		}
+
+		proposalTxs := defaultRes.Txs
+		if injection != nil {
+			proposalTxs = append([][]byte{injection}, proposalTxs...)
+		}
+		return &abci.ResponsePrepareProposal{
+			Txs: proposalTxs,
+		}, nil
+	}
+}
+
+// ProcessProposalHandler handles the ProcessProposal ABCI to validate
+// the canonical set of vote extensions injected by the proposer.
+func (h *Handlers) ProcessProposalHandler() sdk.ProcessProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestProcessProposal) (*abci.ResponseProcessProposal, error) {
+		if req.Height > ctx.ConsensusParams().Abci.VoteExtensionsEnableHeight {
+			var extendedVotes abci.ExtendedCommitInfo
+			if err := json.Unmarshal(req.Txs[0], &extendedVotes); err != nil {
+				h.logger.Error("failed to decode injected extended votes tx", "err", err)
+				return &abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_REJECT}, nil
+			}
+
+			// Validate signatures and perform 2/3 voting power check.
+			err := baseapp.ValidateVoteExtensions(ctx, h.stakingKeeper, req.Height, ctx.ChainID(), extendedVotes)
+			if err != nil {
+				return nil, err
+			}
+
+			// Verify every batch signature.
+			batch, err := h.batchingKeeper.GetBatchForHeight(ctx, ctx.BlockHeight()+proposalOffset)
+			if err != nil {
+				return nil, err
+			}
+			for _, vote := range extendedVotes.Votes {
+				if len(vote.VoteExtension) != 64 {
+					h.logger.Error("invalid length vote extension in proposal", "vote", vote)
+					return nil, ErrInvalidVoteExtensionLength
+				}
+
+				ok, err := h.verifyBatchSignature(ctx, batch.BatchId, vote.VoteExtension[:64], vote.Validator.Address, utils.Secp256k1)
+				if !ok || err != nil {
+					return nil, err
```

Comment:
For the `!ok` case won't the SDK be confused it it receives a `nil` value for both the `ResponseProcessProposal` and the error?

Reply:
Yeah good point. This was a bad mistake. Added a new error `ErrInvalidBatchSignature` to address this.

On a different but related note, I am still confused by how Cosmos is handling errors from these ABCI methods. I find this issue particularly confusing https://github.com/cosmos/cosmos-sdk/pull/17930. Gotta look into it more and test things out.